### PR TITLE
add dp up - down minimum cost for tickets

### PR DESCRIPTION
**Pull request by alexpantyukhin**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/minimum_tickets_cost.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    >>> mincost_tickets([1,2,3,4,5,6,7,8,9,10,30,31], [2,90,150])
+    24
+    """
+
+    from functools import lru_cache
+
+    days_set = set(days)
+
+    @lru_cache(maxsize=None)
+    def dp(index: int) -> int:
```
[the remark this answers is not in the document]
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py:42]: This is an inner function. It's tested by outer levels tests.

File: dynamic_programming/minimum_tickets_cost.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def mincost_tickets(days: list[int], costs: list[int]) -> int:
+    """
+    >>> mincost_tickets([1,4,6,7,8,20], [2,7,15])
+    11
+    >>> mincost_tickets([1,2,3,4,5,6,7,8,9,10,30,31], [2,7,15])
+    17
+    >>> mincost_tickets([1,2,3,4,5,6,7,8,9,10,30,31], [2,90,150])
+    24
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:34]: Please add tests with zero, negative numbers, floating point numbers, and mixing empty lists and non-empty lists.

File: dynamic_programming/minimum_tickets_cost.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    """
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:35]:
```suggestion
    >>> dp(366)
    0
    """
```

File: dynamic_programming/minimum_tickets_cost.py (hunk 2 above)
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py:36]: added tests for different cases. The function expects integers values only no floats. Should we add anyway float tests anyway?

File: dynamic_programming/minimum_tickets_cost.py (hunk 3 above)
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py:49]: actually the dp function is inside scope. Should we add this?

File: dynamic_programming/minimum_tickets_cost.py (hunk 2 above)
Comment (cclauss) [reply dynamic_programming/minimum_tickets_cost.py:36]: Yes.  The caller can throw any garbage they want at your function.  Letters, dicts, sets, None so be ready.

Type hints are only hints and callers can do what they want.  This is called Negative User Testing.

File: dynamic_programming/minimum_tickets_cost.py (hunk 3 above)
Comment (cclauss) [reply dynamic_programming/minimum_tickets_cost.py:57]: If it works then add it.  Just because it is an inner function does not mean that it is bug free.

File: dynamic_programming/minimum_tickets_cost.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Runtime complexity: O(n)
+
+The implementation was tested on the
+leetcode: https://leetcode.com/problems/minimum-cost-for-tickets/
+"""
+
+from typing import List
+
+"""
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:23]:
```suggestion

```

File: dynamic_programming/minimum_tickets_cost.py (hunk 2 above)
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py:31]: fixed

File: dynamic_programming/minimum_tickets_cost.py (hunk 3 above)
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py:52]: fixed

File: dynamic_programming/minimum_tickets_cost.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    >>> mincost_tickets([2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [1, 2, 3, 4])
+    Traceback (most recent call last):
+     ...
+    ValueError: The lengths of costs should be equal 3
+    """
+
+    # Validation
+    if not isinstance(days, list):
+        raise ValueError("The parameter days should be a list")
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:91]:
```suggestion
    if not isinstance(days, list) or not all(isinstance(day, int) for day in days):
        raise ValueError("The parameter days should be a list of integers")
```

File: dynamic_programming/minimum_tickets_cost.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    if not isinstance(costs, list):
+        raise ValueError("The parameter costs should be a list")
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:94]:
```suggestion
    if len(costs) !=3 or not all(isinstance(cost, int) for cost in costs):
        raise ValueError("The parameter costs should be a list of three integers")
```

File: dynamic_programming/minimum_tickets_cost.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    if len(days) == 0:
+        return 0
+
+    if min(days) <= 0:
+        raise ValueError("All days elements should be greater than 0")
+
+    if max(days) >= 366:
+        raise ValueError("All days elements should be less than 366")
+
+    from functools import lru_cache
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:114]: Please import at a global scope (at the top of the file).

**Review by cclauss — CHANGES_REQUESTED** [review]
Looking great.  Please consolidate error checking and import at the top of the file and then we can land this one.

File: dynamic_programming/minimum_tickets_cost.py (hunk 7 above)
Comment (alexpantyukhin) [reply dynamic_programming/minimum_tickets_cost.py]: fixed

File: dynamic_programming/minimum_tickets_cost.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    if any([not isinstance(day, int) for day in days]):
+        raise ValueError("All days elements should be integer values")
+
+    if any([not isinstance(cost, int) for cost in costs]):
+        raise ValueError("All costs elements should be integer values")
+
+    if len(costs) != 3:
+        raise ValueError("The lengths of costs should be equal 3")
+
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:106]: No longer required...
```suggestion
```

File: dynamic_programming/minimum_tickets_cost.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        if index not in days_set:
+            return dp(index + 1)
+
+        return min(
+            costs[0] + dp(index + 1),
+            costs[1] + dp(index + 7),
+            costs[2] + dp(index + 30),
+        )
+
+    return dp(1)
```
Comment (cclauss) [inline dynamic_programming/minimum_tickets_cost.py:128]:
```suggestion
    def dynamic_programming(index: int) -> int:
        if index > 365:
            return 0

        if index not in days_set:
            return dp(index + 1)

        return min(
            costs[0] + dp(index + 1),
            costs[1] + dp(index + 7),
            costs[2] + dp(index + 30),
        )

    return dynamic_programming(1)
```

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
Thanks for this!
